### ISSUE TITLE
typing.Self inside dataclass and Pydantic model field causes false positive invalid-argument-type

### ISSUE THREAD

**Issue by danila-schelkov** [astral-sh/ty#2608]
### Summary

typing.Self behaves differently in dataclass fields compared to regular class initializers.

```py
from dataclasses import dataclass
from typing import Self


@dataclass
class Node:
    children: list[Self]


_node = Node(children=[Node([])])
```

```
> ty check test.py
error[invalid-argument-type]: Argument is incorrect
  --> test.py:10:14
   |
10 | _node = Node(children=[Node([])])
   |              ^^^^^^^^^^^^^^^^^^^ Expected `list[<special-form 'typing.Self'>]`, found `list[<special-form 'typing.Self'> | Node]`
   |
info: rule `invalid-argument-type` is enabled by default

Found 1 diagnostic
```

Meanwhile, the same annotation works correctly when used in a regular \_\_init__ method:

```py
from typing import Self


class Node:
    def __init__(self, children: list[Self]) -> None:
        self.children: list[Self] = children


_node = Node(children=[Node([])])

```

```
> ty check test.py
All checks passed!
```

### Version

ty 0.0.13 (fc1478bd9 2026-01-21)

**Comment by charliermarsh** [comment]
Thanks! This would be fixed by https://github.com/astral-sh/ruff/pull/22487 and is the same as https://github.com/astral-sh/ty/issues/1124.
